### DISCUSSION TITLE
Feature Requests

### DISCUSSION THREAD

**Discussion by krzyzanowskim** [krzyzanowskim/STTextView#14] [Questions]
If you're looking for a new feature, feel free to suggest it or **vote** for the one already proposed below!

(want to sponsor a new feature from the list or outside the list? contact me)

**Comment by krzyzanowskim** [comment]
**Source Code highlighting support**

Use tree-sitter parser-based parsers to highlight text content

**Reply by krzyzanowskim** [reply]
While it's fairly easy to add, it brings quite a bit of dependencies that are mostly a burden for just a TextView component. I'm thinking about the way to avoid requirement of the tree-sitter dependencies (make it optional)

**Reply by Mx-Iris** [reply]
Please consider making this optional. I'm using sourcekitd to implement syntax highlighting

**Reply by krzyzanowskim** [reply]
Recently I added **Plugins** to STTextView, and one of the first plugins is [Plugin-Neon](https://github.com/krzyzanowskim/STTextView-Plugin-Neon) Syntax Highlighting. Plugins are optional, separate pieces that can be added to STTextView instance at will.

This is still in the early stage (need `main` branch of STTextView).

I encourage everyone interested in the feature to contribute to the plugin, especially in these areas:

- Documentation
- Theming/coloring
- Configure supported syntaxes
- Overall fixes and improvements

**Comment by krzyzanowskim** [comment]
**Markers**

visual line markers on the line number ruler (`STLineNumberRulerView`)

**Comment by krzyzanowskim** [comment]
*Full multi-cursor editing support*

There's already some support for that feature but it needs more work to feel complete. I found the VSCode approach appealing so that would be a reference (not necessarily how Xcode works)

The areas to improvement are:
- Visible multiple cursors
- Adding new cursors with the ~Option key (akin to VSCode)~ Shift + Control (akin to Xcode)

**Reply by nighthawk** [reply]
One of my main uses of multiple cursors is editing the same thing across its various occurrences in one go, using the "select next occurrence" option (Cmd+Opt+E in Xcode or Cmd+D in VSCode). 

One thing that bugs me about Xcode there is that it doesn't handle the case well where the next occurrence is right after the first. VSCode handles it fine. Example:

You have the text `https://github.com/krzyzanowskim/STTextView` and you want to escape the forward slashes. You'd expect to select the first slash, then press the "select next occurrence" command, until all slashes are selected and then type left arrow and then backslash. This works in VSCode but doesn't work in Xcode. (On the first "select next occurrence" Xcode will not actually create two cursors both selecting a slash, but instead just has one selection selecting both slashes.) Xcode has the same issue if you want to un-escape `https:\/\/github.com\/krzyzanowskim\/STTextView`, too.

**Comment by chriseidhof** [comment]
Have you considered a markdown mode? This could also be a separate (paid?) component built on top of STTextView. I definitely had projects where I would have paid for this / sponsor this.

Also, is there support for custom text storages, and if yes, how does that work? I've had a niche need for this (I was trying to build a CRDT that records the exact edits so I could replay them later).

**Reply by krzyzanowskim** [reply]
> Have you considered a markdown mode

Yes, however as of now I think about Markdown highlight via https://github.com/krzyzanowskim/STTextView/discussions/14#discussioncomment-4917394 that would look like this:

<img width="302" alt="Screenshot 2023-02-09 at 14 32 30" src="https://user-images.githubusercontent.com/758033/217826646-95f2e1c0-4fa1-4ff0-a01b-f6817281261a.png">

more advanced editing with in-place replacement of markup is possible, yet further on my priority as of today

**Reply by krzyzanowskim** [reply]
> is there support for custom text storages

yes. and not out of the box

Yes because it can use any TextKit 2 storage.
Not out of the box as I rely on NSTextStorage for attributed string and content manipulation because that part is broken in TextKit 2 interface (_FB9907261: NSTextElementProvider.replaceContents(in:with:) does not replace content as documented_ and _FB9692714: Rendering attributes does not draw properly_). If custom storage would provide these, then it very much could be used. I'd love to play with CRDT storage!

**Comment by iliasaz** [comment]
Would it be possible to expose the current selection as Binding<Range<String.Index>> over $text so that a SwiftUI wrapper view or MV can easily access it? It could also be used as the cursor position when the range start = end. For multi-cursor/multi-select use case, it could be a Binding<[Range<String.Index>]>.
Thanks!

**Reply by krzyzanowskim** [reply]
In 0.8.7 I exposed a version of selection binding (single selection) as Binding<NSRange>

https://github.com/krzyzanowskim/STTextView/assets/758033/cb79eac4-1045-4dc6-9faa-1c4f9c218c03

**Comment by florianpircher** [comment]
**Access “Spelling & Grammar” functionality using only the keyboard.**

In most Mac apps, to correct a spelling mistake, the mouse pointer needs to be involved, either to bring up the context menu for the underlined word or to pick a correction option in the system “Spelling & Grammar” window.

The editor TextMate has a custom implementation where _Edit_ → _Spelling_ → _Show Spelling and Grammar_ (bound to “⌘;”) selects the next misspelled word and opens a menu with items for replacement options as well as items for ignoring or learning the word. All of these options are also accessible using keyboard shortcuts.

https://github.com/krzyzanowskim/STTextView/assets/3776198/91ceaab5-f40d-4762-afe4-75071cd78d5a

Being able to keep the hands on the keyboard while typing is very nice indeed, and others agree ([The Talk Show: 312: _‘Not to Get Zealotrous’, With Craig Mod_](https://daringfireball.net/thetalkshow/2021/04/11/ep-312), 1:54:10–1:54:58).

**Comment by markmunz** [comment]
**Show Invisibles option**

NSTextView, via NSLayoutManager has an option for showsInvisibleCharacters, which would show whitespace and other "invisible" characters with special glyphs (and we you could tap into). Admittedly, OS substitutions weren't great, but you could tap into the layout manager to create invisibles that user would expect (dots for spaces, arrow for tabs, return etc). 

As of macOS 14.0 NSTextLayoutManager (TextKit 2) is still missing that. And honestly, not even sure where to tap into in order to render the invisibles.

**Reply by krzyzanowskim** [reply]
I have one idea of how to do it. I just added `STTextView.showsInvisibleCharacters` (on main branch). It's not full-featured yet, as it only draws dot, but that's a start. If you want to fiddle with it, search for `drawInvisibles`

https://github.com/krzyzanowskim/STTextView/assets/758033/8f76e0c5-2935-4a5d-ab6c-12bef5a8a07b

**Reply by austincondiff** [reply]
@krzyzanowskim You might check out our issue here. We are hoping to allow users to be able to control each type of invisible rendered (space, tab, new line, etc.), both the visibility of each and the character rendered. Great work so far!
- https://github.com/CodeEditApp/CodeEditTextView/issues/207

**Comment by bardigolriz** [comment]
**Ability to have multiple dynamically expanding STTextViews vertically stacked together with _one_ scrollbar**

I would like the ability to have a vertical `ScrollView` whose contents are two or more `STTextView`:

1) Each `STTextView` would expand to fit so its contents is fully visible always (preferably with a `minHeight`).
2) As a result of 1), _no_ `STTextView` has their own their scrollbar.
3) Instead the parent container `ScrollView` is exclusively responsible for handling scrolling.

At the moment, the following code:

       Form {
            STTextViewUI.TextView(
                text: $text,
                selection: $selection,
                options: [.wrapLines]
            ).textViewFont(.monospacedDigitSystemFont(ofSize: NSFont.systemFontSize, weight: .regular))
            
            STTextViewUI.TextView(
                text: $text2,
                selection: $selection2,
                options: [.wrapLines]
            ).textViewFont(.monospacedDigitSystemFont(ofSize: NSFont.systemFontSize, weight: .regular))
        }

Has the two controls vertically stacked taking up 50% of the view, and having individual scrollbars for each. 

Admittedly, this may not be relevant for coding applications, which is possibly a primary use for `STTextView`, but the flexibility this could have broad appeal. To be clear, my personal use case is the ability to have a screen where you can create a thread of messages for social media (Mastodon). I’d like for the `STTextView` to have a `minHeight` of ~5 or so lines visible that expands if the content dictates. There’s a `Button` somewhere to add another message to the thread (`View`). Doing so, would append a `STTextView` below the previous last and so on.

For reference/comparison, the combination of `ScrollView` and `TextEditor` partially works but there are a few noteworthy issues.

1) there’s a flashing/flicker when you start a new line

2) when the caret starts a new line on the last visible line of window, it’s no longer in view 

3) scrolling seems to make the last line of an expanded view to become scrolled out of view too

https://github.com/krzyzanowskim/STTextView/assets/8938930/e8a29054-6404-4aad-877a-592a6f97938b


```
struct ContentView: View {
    
    @State var text: String = ""
    
    @State var text2: String = ""
    
    @State var text3: String = ""
    
    var body: some View {
        ScrollView(.vertical) {
            TextEditor(text: $text)
                .scrollDisabled(true)
                .fixedSize(horizontal: false, vertical: true)
            TextEditor(text: $text2)
                .scrollDisabled(true)
                .fixedSize(horizontal: false, vertical: true)
            TextEditor(text: $text3)
                .scrollDisabled(true)
                .fixedSize(horizontal: false, vertical: true)
        }
    }
}
```

Firstly, hope this is clear. Secondly, hope this is possible. Finally, hope this can be implemented (:

**Comment by ladvoc** [comment]
I see the Neon plug-in depends on `tree-sitter-xcframework` which bundles language support for twelve languages. Is it possible to add support for another language that is not covered by this package?

**Reply by krzyzanowskim** [reply]
yes. It’s changed now an multiple news languages has been added just recently

**Comment by HassanTaleb90** [comment]
Placeholder support

```
let <#name#> = <#value#>
```
<img width="143" alt="Screenshot 2024-08-10 at 11 04 20 PM" src="https://github.com/user-attachments/assets/3dcbed83-76ef-4c68-a4b4-ef86bd980dbe">

**Reply by krzyzanowskim** [reply]
I need that. That is specific to code editor and it feels like if that could be a plugin maybe

**Comment by disc0infern0** [comment]
It's quite criminal that Apple hasn't added any support for Tables yet in any of their text editing views.
Would it be possible to add support into STTextView for Tables?  (perhaps via NSTableView? )
From some google searching, it seems that tables could be added to a textview attachment, but that solution is a) old, and b) something of a hack.

**Reply by colbyn** [reply]
Hey @disc0infern0 

I've actually been working on a rich text view that's designed to support tables, among other things, in otherwise plain text views (drawn using TextKit 2 fragments) as well as supporting all block types from the Github Flavored Markdown Spec. 

Although while the results aren't yet ready for public consumption, here's my [proof of concept prototype project]( https://github.com/SuperSwiftMarkup/SuperSwiftMarkdownPrototype ) that demonstrates this functionality.

**Reply by krzyzanowskim** [reply]
I haven't prioritized tables yet because of the editing. It sounds like a task for a long, lonely winter night.

**Reply by colbyn** [reply]
Oh yeah I should have clarified, I'm currently focused on read only (but selectable) functionality which in some ways I kinda enjoy since I can just focus on aesthetic appeal.

For context I originally started developing this for a chatbot concept based on a branching conversation data model, and the UI tree was such that I could not get vertically stacked web views (nested in stacked tab views for each branch that were nested in a scroll view) to layout without problems... Hence this project to display read-only but selectable markdown formatted content.  

----

I am curious how others would solve this problem for editing behavior since my instinct is to use lossless ASTs which I've been experimenting with but I don't think I've seen anyone do this out in the wild...

Since while text editors generally need to support every format under the sun, if your editor is only going to support markdown and some extensions you can do things that aren't possible with general purpose text buffers and stupid regex based syntax highlighting. For instance this is from a toy markup language for taking school notes. While it's just a proof of concept you can see how I'm able to color things in a much more context sensitive manner such as coloring blocks based on nesting level as shown in the following screenshots (I really wish I took the time to prettify this a bit more since I actually think this idea its super cool): 

![Screenshot 2023-11-03 at 11 04 54 PM](https://github.com/user-attachments/assets/8b6594d4-f468-434b-ba1d-64edafe4ed1e)
![Screenshot 2023-11-03 at 11 20 12 PM](https://github.com/user-attachments/assets/8a6de0f4-d611-42b7-bf21-65023c8c618a)

Although I suppose this would better apply to HTML/LaTeX nested within markdown which is on my TODO list. IMO this makes TeX-like syntax work, it's a shame that people are so against it as a markup language since it has other benefits. 

But I digress...

**Comment by lucasjbyoung** [comment]
Make typingAttributes setter public (currently internal(set))

`typingAttributes` is currently declared as public internal(set), which means external consumers can read but not write it. The README shows textView.typingAttributes[.paragraphStyle] = paragraph as the way to set the default paragraph style, but this doesn't compile because the subscript setter requires write access to the property.

Use case: In a screenplay editor, different lines have different paragraph indents (action at 0, character at 19ch, dialogue at 10ch). When the cursor is on an empty line at the end of the document, the cursor's horizontal position should reflect the expected element type's indent. defaultParagraphStyle only affects the fallback for attribute merging - it doesn't control the cursor position on the extra line fragment. The only way to position the cursor correctly before the user types is to set typingAttributes[.paragraphStyle] with the desired indent, but the internal(set) access level prevents this.

Current workaround: Using lineFragmentPadding as a left margin, which works but forces all paragraph style indents to be relative to the padding rather than absolute - making the formatting model less clear.

Requested change: Change public internal(set) var typingAttributes to public var typingAttributes (or add a public setTypingAttributes(_:) method).

Environment: STTextView 2.3.6, macOS 26, Swift 6

Cheers!

**Reply by krzyzanowskim** [reply]
> README shows textView.typingAttributes[.paragraphStyle] = paragraph 

that's "typo", should be `textView.defaultParagraphStyle = paragraph`

**Reply by lucasjbyoung** [reply]
Thanks for the reply! We actually needed per-line paragraph styles for a screenplay editor (different indents for character names, dialogue, action, etc.). defaultParagraphStyle only sets one style for the whole document. We needed typingAttributes to control the paragraph style of the next typed character when the user switches element types via keyboard shortcuts.

**Reply by krzyzanowskim** [reply]
I understand. I'm trying to remember why did I restrict it in the first place. It seems ok to make it editable, just keep in mind it's not very persistent value and some operations updates it automatically.

I changed it on main branch
